src/main/java/hudson/remoting/JarCacheSupport.java

```diff
@@ -50,50 +51,92 @@
 
         while (true) {// might have to try a few times before we get successfully resolve
 
-            final Checksum key = new Checksum(sum1,sum2);
-            final AsyncFutureImpl<URL> promise = new AsyncFutureImpl<URL>();
-            Future<URL> cur = inprogress.putIfAbsent(key, promise);
+            final Checksum key = new Checksum(sum1,sum2);        
+            Future<URL> cur = inprogress.get(key);
             if (cur!=null) {
                 // this computation is already in progress. piggy back on that one
                 return cur;
             } else {
                 // we are going to resolve this ourselves and publish the result in 'promise' for others
-                downloader.submit(new Runnable() {
-                    public void run() {
-                        try {
-                            URL url = retrieve(channel,sum1,sum2);
-                            inprogress.remove(key);
-                            promise.set(url);
-                        } catch (ChannelClosedException e) {
-                            // the connection was killed while we were still resolving the file
-                            bailout(e);
-                        } catch (RequestAbortedException e) {
-                            // the connection was killed while we were still resolving the file
-                            bailout(e);
-                        } catch (InterruptedException e) {
-                            // we are bailing out, but we need to allow another thread to retry later.
-                            bailout(e);
+                try {
+                    final AsyncFutureImpl<URL> promise = new AsyncFutureImpl<URL>();
+                    ExecutorServiceUtils.submitAsync(downloader, new  DownloadRunnable(channel, sum1, sum2, key, promise));
+                    // Now we are sure that the task has been accepted to the queue, hence we cache the promise
+                    // if nobody else caches it before.
+                    inprogress.putIfAbsent(key, promise);
```

Comment:
With this logic I am not fully convinced that `ConcurrentMap` is still a good idea here, maybe a common lock would be preferable since it allows to avoid duplication on race conditions and then the deduplication logic.

In this case there is also a risk that `DownloadRunnable` gets executed before putting the promise into the cache. 
And yes, putIfAbsent here should return the value if something is returned.